crates/storage/provider/src/post_state.rs

```diff
@@ -0,0 +1,704 @@
+            });
+
+        // Write account changes
+        let mut account_changeset_cursor = tx.cursor_dup_write::<tables::AccountChangeSet>()?;
+        for changeset in account_changes.into_iter() {
+            match changeset {
+                Change::AccountDestroyed { id, address, old } |
+                Change::AccountChanged { id, address, old, .. } => {
+                    account_changeset_cursor.append_dup(
+                        first_transition_id + id,
+                        AccountBeforeTx { address, info: Some(old) },
+                    )?;
+                }
+                Change::AccountCreated { id, address, .. } => {
+                    account_changeset_cursor.append_dup(
+                        first_transition_id + id,
+                        AccountBeforeTx { address, info: None },
+                    )?;
+                }
+                _ => unreachable!(),
+            }
+        }
+
+        // Write storage changes
+        let mut storages_cursor = tx.cursor_dup_write::<tables::PlainStorageState>()?;
+        let mut storage_changeset_cursor = tx.cursor_dup_write::<tables::StorageChangeSet>()?;
+        for changeset in storage_changes.into_iter() {
+            match changeset {
+                Change::StorageChanged { id, address, changeset } => {
+                    let storage_id = TransitionIdAddress((first_transition_id + id, address));
+
+                    for (key, (old_value, _)) in changeset {
+                        storage_changeset_cursor.append_dup(
+                            storage_id,
+                            StorageEntry { key: H256(key.to_be_bytes()), value: old_value },
+                        )?;
+                    }
+                }
+                Change::StorageWiped { id, address } => {
+                    let storage_id = TransitionIdAddress((first_transition_id + id, address));
+
+                    if let Some((_, entry)) = storages_cursor.seek_exact(address)? {
+                        storage_changeset_cursor.append_dup(storage_id, entry)?;
+
+                        while let Some(entry) = storages_cursor.next_dup_val()? {
+                            storage_changeset_cursor.append_dup(storage_id, entry)?;
+                        }
+                    }
+                }
+                _ => unreachable!(),
+            }
+        }
+
+        // Write new storage state
+        for (address, storage) in self.storage.into_iter() {
+            if storage.wiped {
+                if storages_cursor.seek_exact(address)?.is_some() {
+                    storages_cursor.delete_current_duplicates()?;
+                }
+
+                // If the storage is marked as wiped, it might still contain values. This is to
+                // avoid deallocating where possible, but these values should not be written to the
+                // database.
+                continue
+            }
+
+            for (key, value) in storage.storage {
```

Comment:
wondering if we could try and check if this address has any storage set (`first_dup().is_some()`), and if not, use `append` instead of `upsert`.

Reply:
if we can associate these changes with an `AccountCreated` (as long as there has been no `AccountDestroyed`), that would avoid querying the `first_dup` .

I'm not just thinking about speed, but also table size. `append` leads to smaller table size. i'm assuming the same happens with `append_dup`, but maybe it's not worth the added complexity

Reply:
I tried this locally and it made sync 30% slower, I don't think it's worth it.